prow/tide/status.go

```diff
@@ -547,46 +559,93 @@ func (sc *statusController) sync(pool map[string]PullRequest, blocks blockers.Bl
 }
 
 func (sc *statusController) search() []PullRequest {
-	queries := sc.config().Tide.Queries
-	if len(queries) == 0 {
+	rawQueries := sc.config().Tide.Queries
+	if len(rawQueries) == 0 {
 		return nil
 	}
 
-	orgExceptions, repos := queries.OrgExceptionsAndRepos()
+	orgExceptions, repos := rawQueries.OrgExceptionsAndRepos()
 	orgs := sets.StringKeySet(orgExceptions)
-	query := openPRsQuery(orgs.List(), repos.List(), orgExceptions)
-	now := time.Now()
-	log := sc.logger.WithField("query", query)
-	if query != sc.PreviousQuery {
-		// Query changed and/or tide restarted, recompute everything
-		log.WithField("previously", sc.PreviousQuery).Info("Query changed, resetting start time to zero")
-		sc.LatestPR = metav1.Time{}
-		sc.PreviousQuery = query
-	}
-
-	// TODO @alvaroaleman: Add github apps support
-	prs, err := search(sc.ghc.QueryWithGitHubAppsSupport, sc.logger, query, sc.LatestPR.Time, now, "")
-	log.WithField("duration", time.Since(now).String()).Debugf("Found %d open PRs.", len(prs))
+	queries := openPRsQueries(orgs.List(), repos.List(), orgExceptions)
+	if !sc.usesGitHubAppsAuth {
+		var query string
+		for _, orgQuery := range queries {
+			query += " " + orgQuery
+		}
+		queries = map[string]string{"": query}
+	}
+
+	if sc.storedState == nil {
+		sc.storedState = map[string]storedState{}
+	}
+
+	var prs []PullRequest
+	var errs []error
+	var lock sync.Mutex
+	var wg sync.WaitGroup
+
+	for org, query := range queries {
+		org, query := org, query
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+			now := time.Now()
+			log := sc.logger.WithField("query", query)
+
+			sc.storedStateLock.Lock()
+			latestPR := sc.storedState[org].LatestPR
+			if query != sc.storedState[org].PreviousQuery {
+				// Query changed and/or tide restarted, recompute everything
+				log.WithField("previously", sc.storedState[org].PreviousQuery).Info("Query changed, resetting start time to zero")
+				sc.storedState[org] = storedState{PreviousQuery: query}
+			}
+			sc.storedStateLock.Unlock()
+
+			result, err := search(sc.ghc.QueryWithGitHubAppsSupport, sc.logger, query, latestPR.Time, now, org)
+			log.WithField("duration", time.Since(now).String()).WithField("result_count", len(result)).Debug("Searched for open PRs.")
+
+			func() {
+				sc.storedStateLock.Lock()
+				defer sc.storedStateLock.Unlock()
+
+				log := log.WithField("latestPR", sc.storedState[org].LatestPR)
+				if len(result) == 0 {
+					log.Debug("no new results")
```

Comment:
Looks like a good candidate for a `return` to allow shifting left the code below? (L618 too)

Reply:
Good point, updated